File: internal/executor/mock.go
package executor

import (
	"context"
	"fmt"
	"strings"
	"sync"
)

// RecordedCommand is a command captured by MockExecutor.
type RecordedCommand struct {
	Cmd   []string
	Sudo  bool
	Shell bool
}

// MockExecutor records commands without executing them.
// It can return pre-configured responses for specific command patterns.
type MockExecutor struct {
	mu        sync.Mutex
	Commands  []RecordedCommand
	Files     map[string]string
	responses map[string]*CommandResult
}

// NewMockExecutor creates a MockExecutor ready for use.
func NewMockExecutor() *MockExecutor {
	return &MockExecutor{
		Files:     make(map[string]string),
		responses: make(map[string]*CommandResult),
	}
}

// SetResponse pre-configures a response for a command key (space-joined args).
func (m *MockExecutor) SetResponse(cmdKey string, result *CommandResult) {
	m.mu.Lock()
	defer m.mu.Unlock()
	m.responses[cmdKey] = result
}

func (m *MockExecutor) getResponse(cmd []string) *CommandResult {
	key := strings.Join(cmd, " ")
	if r, ok := m.responses[key]; ok {
		return r
	}
	return &CommandResult{Success: true, ReturnCode: 0, Stdout: "", Command: cmd}
}

func (m *MockExecutor) IsDryRun() bool {
	return false
}

func (m *MockExecutor) Execute(_ context.Context, cmd []string) (*CommandResult, error) {
	m.mu.Lock()
	defer m.mu.Unlock()
	m.Commands = append(m.Commands, RecordedCommand{Cmd: cmd, Sudo: false})
	return m.getResponse(cmd), nil
}

func (m *MockExecutor) ExecuteSudo(_ context.Context, cmd []string) (*CommandResult, error) {
	m.mu.Lock()
	defer m.mu.Unlock()
	m.Commands = append(m.Commands, RecordedCommand{Cmd: cmd, Sudo: true})
	return m.getResponse(cmd), nil
}

func (m *MockExecutor) ExecuteShell(_ context.Context, cmd string) (*CommandResult, error) {
	m.mu.Lock()
	defer m.mu.Unlock()
	parts := []string{"sh", "-c", cmd}
	m.Commands = append(m.Commands, RecordedCommand{Cmd: parts, Shell: true})
	return m.getResponse(parts), nil
}

func (m *MockExecutor) FileExists(path string) bool {
	m.mu.Lock()
	defer m.mu.Unlock()
	_, ok := m.Files[path]
	return ok
}

func (m *MockExecutor) ReadFile(path string) (string, error) {
	m.mu.Lock()
	defer m.mu.Unlock()
	content, ok := m.Files[path]
	if !ok {
		return "", fmt.Errorf("mock file not found: %s", path)
	}
	return content, nil
}

func (m *MockExecutor) WriteFile(path string, content string, sudo bool) (*CommandResult, error) {
	m.mu.Lock()
	defer m.mu.Unlock()
	m.Files[path] = content
	cmd := []string{"write", path}
	m.Commands = append(m.Commands, RecordedCommand{Cmd: cmd, Sudo: sudo})
	return &CommandResult{Success: true, ReturnCode: 0, Command: cmd}, nil
}

// CommandCount returns the number of recorded commands.
func (m *MockExecutor) CommandCount() int {
	m.mu.Lock()
	defer m.mu.Unlock()
	return len(m.Commands)
}

// HasCommand checks if a command matching the given prefix was recorded.
func (m *MockExecutor) HasCommand(prefix ...string) bool {
	m.mu.Lock()
	defer m.mu.Unlock()
	for _, rec := range m.Commands {
		if len(rec.Cmd) >= len(prefix) {
			match := true
			for i, p := range prefix {
				if rec.Cmd[i] != p {
					match = false
					break
				}
			}
			if match {
				return true
			}
		}
	}
	return false
}

File: src/cask/executor/mock.py
"""Mock executor for testing."""
from __future__ import annotations

from dataclasses import dataclass, field

from cask.result import ExecResult


@dataclass
class _MockResponse:
    cmd_prefix: str
    exit_code: int
    stdout: str


class MockExecutor:
    """Records calls and returns preconfigured responses."""

    def __init__(self, dry_run: bool = False) -> None:
        self.dry_run = dry_run
        self._calls: list[str] = []
        self._files: dict[str, str] = {}
        self._responses: list[_MockResponse] = []

    @property
    def call_count(self) -> int:
        return len(self._calls)

    @property
    def calls(self) -> list[str]:
        return list(self._calls)

    def was_called(self, substring: str) -> bool:
        return any(substring in c for c in self._calls)

    def call_at(self, index: int) -> str:
        return self._calls[index]

    def add_response(self, cmd_prefix: str, exit_code: int = 0, stdout: str = "") -> None:
        self._responses.append(_MockResponse(cmd_prefix, exit_code, stdout))

    def _find_response(self, cmd: list[str]) -> ExecResult:
        cmd_str = cmd[0] if cmd else ""
        for resp in self._responses:
            if cmd_str.startswith(resp.cmd_prefix):
                return ExecResult(resp.exit_code, resp.stdout, "")
        return ExecResult(0, "", "")

    def _record(self, prefix: str | None, cmd: list[str]) -> None:
        parts = []
        if prefix:
            parts.append(prefix)
        parts.extend(cmd)
        self._calls.append(" ".join(parts))

    async def execute(self, cmd: list[str]) -> ExecResult:
        self._record(None, cmd)
        return self._find_response(cmd)

    async def execute_sudo(self, cmd: list[str]) -> ExecResult:
        self._record("[sudo]", cmd)
        return self._find_response(cmd)

    async def execute_shell(self, cmd: str) -> ExecResult:
        self._record("[shell]", [cmd])
        return self._find_response([cmd])

    def file_exists(self, path: str) -> bool:
        return path in self._files

    def read_file(self, path: str) -> str:
        return self._files.get(path, "")

    def write_file(self, path: str, content: str, sudo: bool = False) -> None:
        self._files[path] = content
